text/0000-infallible-promotion.md

```diff
@@ -0,0 +1,193 @@
+Notice that this is a problem only for implicit promotion, because with explicit promotion, the value *has* to be known at compile-time -- so stopping compilation if the value cannot be determined is the right behavior.
+
+To solve this problem, every part of the compiler that works with constants needs to be able to handle the case where the constant *has no defined value*, and continue in some correct way.
+This is hard to get right, and has lead to a number of problems over the years:
+* There has been at least one [soundness issue](https://github.com/rust-lang/rust/issues/50814).
+* There are still outstanding [diagnostic issues](https://github.com/rust-lang/rust/issues/61821).
+* Promotion needs a special [exception in const-value validation](https://github.com/rust-lang/rust/issues/67534).
+* All code handling constants has to carry [extra complexity to support promotion](https://github.com/rust-lang/rust/issues/75461)
+
+This RFC proposes to fix all these problems at once, by restricting implicit promotion to those expression whose evaluation cannot fail.
+This is the last step in a series of changes that have been going on for quite some time, starting with the [introduction](https://github.com/rust-lang/rust/pull/53851) of the `#[rustc_promotable]` attribute to control which function calls may be subject to implicit promotion (the original RFC said that all calls to `const fn` should be promoted, but as user-defined `const fn` got closer and closer, that seemed less and less like a good idea, due to all the ways in which evaluating a `const fn` can fail).
+Together with [some planned changes for evaluation of regular constants](https://github.com/rust-lang/rust/issues/71800), this means that all CTFE failures can be made hard errors, greatly simplifying the parts of the compiler that trigger evaluation of constants and handle the resulting value or error.
+
+For more details, see [the MCP that preceded this RFC](https://github.com/rust-lang/lang-team/issues/58).
+
+[promotion-status]: https://github.com/rust-lang/const-eval/blob/33053bb2c9a0c6a17acd3116dd47bbb360e060db/promotion.md
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+(Based on [RFC 1414][rfc1414])
+
+Inside a function body's block:
+
+- If a shared reference to a constexpr rvalue is taken. (`&<constexpr>`),
+- And the constexpr does not contain a `UnsafeCell { ... }` constructor,
+- And the constexpr only consists of operations that will definitely succeed to
+  evaluate at compile-time,
+- And the resulting value does not need dropping,
+- Then instead of translating the value into a stack slot, translate
+  it into a static memory location and give the resulting reference a
+  `'static` lifetime.
+
+Operations that definitely succeed include:
+- literals of any kind
+- constructors (struct/enum/union/tuple)
+- struct/tuple field accesses
+- arithmetic that does not involve division: `+`/`-`/`*`
```

Comment:
Reducing arithmetic to `+,-,*` could be useful on itself for simplicity.
Bitwise operations are missing here (`^,|,&`) and the negation operators `!,-` as every cycle counts. ;-)
Probably also shifting. Or are `<<` and `>>` magically to arithmetic overflows?

[Types of overflows](https://doc.rust-lang.org/reference/expressions/operator-expr.html#overflow) could be linked, but might be not too useful.

Reply:
The list was just meant to be some examples, but it seems it is impossible to stop people from wanting to make it exhaustive, even if that makes it will inevitably become outdated.^^